### ISSUE TITLE
[question] How to package library with protobuf-generated code, given a generator is also a package

### ISSUE THREAD

**Issue by ngrodzitski** [conan-io/conan#10238]
What would be a good approach to package protobuf-generated code?

I want to create a package for library which would also contain *.pb.h header files and associated object-files (*.pb.cc) would be a part of my lib. And what bothers me is that client (the one who uses a package as a dependency) might use different version of protobuf. And I want to make sure that library is get compiled on the host in case current project uses different version of protobuf (already have a case protobuf 2.6 (custom one) vs 3.*)

Maybe any idiomatic mechanics exists for generated code in general? Because the same case might come with any other generated tool (lib part + generator part) that is also a conan-package, and for different versions might generate different code which is tied to a library part of the tool's specific version.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @ngrodzitski 

Not sure if I fully understand your scenario, so please let me throw some ideas in case it helps, but let me know if otherwise.

I think the concept of the ``package_id`` and how dependencies affect the current package_id and binary is what could make sense here. If you have a given package that depends on ``protobuf`` library package, the default is already to request a new binary for every major version of the dependency. So if your package is called "stubs/1.0" -> "protobuf", then "stubs/1.0" will have 2 different binaries, one when depending on "protobuf/2" and a different one when depending on "protobuf/3".

Also, maybe the fact that generated source code is typically generated on the fly in the ``build()`` method, as necessary (typically having a ``build_require`` to ``protobuf`` to be able to call ``protoc`` easily) could help? The source code of a given package should be the same for every platform, and configuration, depending only the exact version of the package (and revision if used), but source code cannot be different for different configurations.

**Comment by ngrodzitski** [comment]
Hi @memsharded 
Let's say my XXX package uses protobuf and XXX/1.2.3 sources are fixed. XXX/1.2.3 works with protobuf 2.6 and 3.17, the sources are the same. Then I want to use XXX/1.2.3 in project YYY and ZZZ, YYY uses protobuf 2.6 and ZZZ uses protobuf  3.17. Both YYY and ZZZ define protobuf version to be used explicitly.
As I understand from you response and from [this page](https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#using-package-id-for-package-dependencies) the solution would look like this: 

```python
# In XXX conanfile.py
def package_id(self):
    self.info.requires["protobuf"].minor_mode()
```

That would cause the build (when doing `conan install` for YYY or ZZZ) for XXX package that depends on protobuf. And within a YYY or ZZZ If I change its protobuf version from `3.17.1` `3.19.1` and obviously to `2.6.1` (custom one) this would trigger XXX rebuild.

And if I would have MyGeneratorLib (which is lib+gen parts) then to trigger a separate build for a similar XXX lib which would depend on MyGeneratorLib the `package_id` function would look like the following:
```python
# In XXX conanfile.py
def package_id(self):
    self.info.requires["MyGeneratorLibf"].patch_mode()
```

Do I get it right?

**Comment by memsharded** [comment]
> That would cause the build (when doing conan install for YYY or ZZZ) for XXX package that depends on protobuf. And within a YYY or ZZZ If I change its protobuf version from 3.17.1 3.19.1 and obviously to 2.6.1 (custom one) this would trigger XXX rebuild.

Yes, exactly. The evidence is in the ``conan search xxx/1.2.3@`` that will list your binaries. You should see 2 different binaries, with 2 different package_ids, because they have different dependencies, even if the rest of the settings and options are the same.

> And if I would have MyGeneratorLib (which is lib+gen parts) then to trigger a separate build for a similar XXX lib which would depend on MyGeneratorLib the package_id function would look like the following:

Not sure if I get it right, but in case you don't know it, you can also set the global ``default_package_id_mode`` in configuration. If you set it to ``minor_mode``, that also applies transitively to other consumers. So indirect consumers of protobuf, will still get a different package_id when the upstream protobuf version changes (the minor), even if they don't explicitly declare it. Is maybe this the behavior you want?

**Comment by memsharded** [comment]
Closing this ticket as outdated, it was Conan 1.X. Things have changed a lot since then, Conan 2 got a lot of improvements and changes regarding ``package_id`` computation.

Please open new tickets for any further question or issue, thanks for the feedback!
